File: app/src/main/java/es/etologic/mahjongscoring2/domain/model/GameRounds.java
package es.etologic.mahjongscoring2.domain.model;

import java.util.List;

import es.etologic.mahjongscoring2.domain.model.enums.TableWinds;

import static es.etologic.mahjongscoring2.domain.model.enums.TableWinds.EAST;
import static es.etologic.mahjongscoring2.domain.model.enums.TableWinds.NORTH;
import static es.etologic.mahjongscoring2.domain.model.enums.TableWinds.SOUTH;
import static es.etologic.mahjongscoring2.domain.model.enums.TableWinds.WEST;

public class GameRounds {

    public static int getTotalScoreP1(List<Round> rounds) {
        int totalPoints = 0;
        for (Round round : rounds) totalPoints += round.getPointsP1();
        return totalPoints;
    }

    public static int getTotalScoreP2(List<Round> rounds) {
        int totalPoints = 0;
        for (Round round : rounds) totalPoints += round.getPointsP2();
        return totalPoints;
    }

    public static int getTotalScoreP3(List<Round> rounds) {
        int totalPoints = 0;
        for (Round round : rounds) totalPoints += round.getPointsP3();
        return totalPoints;
    }

    public static int getTotalScoreP4(List<Round> rounds) {
        int totalPoints = 0;
        for (Round round : rounds) totalPoints += round.getPointsP4();
        return totalPoints;
    }

    public static TableWinds getEastPlayerCurrentSeat(int roundId) {
        switch (roundId) {
            case 1:
            case 2:
            case 3:
            case 4:
            default:
                return EAST;
            case 5:
            case 6:
            case 7:
            case 8:
                return SOUTH;
            case 9:
            case 10:
            case 11:
            case 12:
                return WEST;
            case 13:
            case 14:
            case 15:
            case 16:
                return NORTH;
        }
    }

    public static TableWinds getSouthSeatPlayerByRound(int roundId) {
        switch (roundId) {
            case 1:
            case 2:
            case 3:
            case 4:
            default:
                return SOUTH;
            case 5:
            case 6:
            case 7:
            case 8:
                return EAST;
            case 9:
            case 10:
            case 11:
            case 12:
                return NORTH;
            case 13:
            case 14:
            case 15:
            case 16:
                return WEST;
        }
    }

    public static TableWinds getWestSeatPlayerByRound(int roundId) {
        switch (roundId) {
            case 1:
            case 2:
            case 3:
            case 4:
            default:
                return WEST;
            case 5:
            case 6:
            case 7:
            case 8:
                return NORTH;
            case 9:
            case 10:
            case 11:
            case 12:
                return SOUTH;
            case 13:
            case 14:
            case 15:
            case 16:
                return EAST;
        }
    }

    public static TableWinds getNorthSeatPlayerByRound(int roundId) {
        switch (roundId) {
            case 1:
            case 2:
            case 3:
            case 4:
            default:
                return NORTH;
            case 5:
            case 6:
            case 7:
            case 8:
                return WEST;
            case 9:
            case 10:
            case 11:
            case 12:
                return EAST;
            case 13:
            case 14:
            case 15:
            case 16:
                return SOUTH;
        }
    }

    public static TableWinds getPlayerInitialSeatByCurrentSeat(TableWinds seatPosition, int roundId) {
        switch (roundId) {
            case 1:
            case 2:
            case 3:
            case 4:
                return getPlayerInitialPositionBySeatInRoundEast(seatPosition);
            case 5:
            case 6:
            case 7:
            case 8:
                return getPlayerInitialPositionBySeatInRoundSouth(seatPosition);
            case 9:
            case 10:
            case 11:
            case 12:
                return getPlayerInitialPositionBySeatInRoundWest(seatPosition);
            case 13:
            case 14:
            case 15:
            case 16:
            default:
                return getPlayerInitialPositionBySeatInRoundNorth(seatPosition);

        }
    }

    private static TableWinds getPlayerInitialPositionBySeatInRoundEast(TableWinds seatPosition) {
        switch (seatPosition) {
            case EAST:
                return EAST;
            case SOUTH:
                return SOUTH;
            case WEST:
                return WEST;
            case NORTH:
            default:
                return NORTH;
        }
    }

    private static TableWinds getPlayerInitialPositionBySeatInRoundSouth(TableWinds seatPosition) {
        switch (seatPosition) {
            case EAST:
                return SOUTH;
            case SOUTH:
                return EAST;
            case WEST:
                return NORTH;
            case NORTH:
            default:
                return WEST;
        }
    }

    private static TableWinds getPlayerInitialPositionBySeatInRoundWest(TableWinds seatPosition) {
        switch (seatPosition) {
            case EAST:
                return WEST;
            case SOUTH:
                return NORTH;
            case WEST:
                return SOUTH;
            case NORTH:
            default:
                return EAST;
        }
    }

    private static TableWinds getPlayerInitialPositionBySeatInRoundNorth(TableWinds seatPosition) {
        switch (seatPosition) {
            case EAST:
                return NORTH;
            case SOUTH:
                return WEST;
            case WEST:
                return EAST;
            case NORTH:
            default:
                return SOUTH;
        }
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/domain/model/GameRounds.kt
package es.etologic.mahjongscoring2.domain.model

import es.etologic.mahjongscoring2.domain.model.enums.TableWinds
import es.etologic.mahjongscoring2.domain.model.enums.TableWinds.*

open class GameRounds {
    
    companion object {
        
        fun getTotalScoreP1(rounds: List<Round>): Int {
            var totalPoints = 0
            for (round in rounds) totalPoints += round.pointsP1
            return totalPoints
        }
        
        fun getTotalScoreP2(rounds: List<Round>): Int {
            var totalPoints = 0
            for (round in rounds) totalPoints += round.pointsP2
            return totalPoints
        }
        
        fun getTotalScoreP3(rounds: List<Round>): Int {
            var totalPoints = 0
            for (round in rounds) totalPoints += round.pointsP3
            return totalPoints
        }
        
        fun getTotalScoreP4(rounds: List<Round>): Int {
            var totalPoints = 0
            for (round in rounds) totalPoints += round.pointsP4
            return totalPoints
        }
        
        fun getEastPlayerCurrentSeat(roundId: Int): TableWinds {
            return when (roundId) {
                1, 2, 3, 4 -> EAST
                5, 6, 7, 8 -> SOUTH
                9, 10, 11, 12 -> WEST
                13, 14, 15, 16 -> NORTH
                else -> EAST
            }
        }
        
        fun getSouthSeatPlayerByRound(roundId: Int): TableWinds {
            return when (roundId) {
                1, 2, 3, 4 -> SOUTH
                5, 6, 7, 8 -> EAST
                9, 10, 11, 12 -> NORTH
                13, 14, 15, 16 -> WEST
                else -> SOUTH
            }
        }
        
        fun getWestSeatPlayerByRound(roundId: Int): TableWinds {
            return when (roundId) {
                1, 2, 3, 4 -> WEST
                5, 6, 7, 8 -> NORTH
                9, 10, 11, 12 -> SOUTH
                13, 14, 15, 16 -> EAST
                else -> WEST
            }
        }
        
        fun getNorthSeatPlayerByRound(roundId: Int): TableWinds {
            return when (roundId) {
                1, 2, 3, 4 -> NORTH
                5, 6, 7, 8 -> WEST
                9, 10, 11, 12 -> EAST
                13, 14, 15, 16 -> SOUTH
                else -> NORTH
            }
        }
        
        fun getPlayerInitialSeatByCurrentSeat(seatPosition: TableWinds, roundId: Int): TableWinds {
            return when (roundId) {
                1, 2, 3, 4 -> getPlayerInitialPositionBySeatInRoundEast(seatPosition)
                5, 6, 7, 8 -> getPlayerInitialPositionBySeatInRoundSouth(seatPosition)
                9, 10, 11, 12 -> getPlayerInitialPositionBySeatInRoundWest(seatPosition)
                13, 14, 15, 16 -> getPlayerInitialPositionBySeatInRoundNorth(seatPosition)
                else -> getPlayerInitialPositionBySeatInRoundNorth(seatPosition)
            }
        }
        
        private fun getPlayerInitialPositionBySeatInRoundEast(seatPosition: TableWinds): TableWinds {
            return when (seatPosition) {
                EAST -> EAST
                SOUTH -> SOUTH
                WEST -> WEST
                NORTH -> NORTH
                else -> NORTH
            }
        }
        
        private fun getPlayerInitialPositionBySeatInRoundSouth(seatPosition: TableWinds): TableWinds {
            return when (seatPosition) {
                EAST -> SOUTH
                SOUTH -> EAST
                WEST -> NORTH
                NORTH -> WEST
                else -> WEST
            }
        }
        
        private fun getPlayerInitialPositionBySeatInRoundWest(seatPosition: TableWinds): TableWinds {
            return when (seatPosition) {
                EAST -> WEST
                SOUTH -> NORTH
                WEST -> SOUTH
                NORTH -> EAST
                else -> EAST
            }
        }
        
        private fun getPlayerInitialPositionBySeatInRoundNorth(seatPosition: TableWinds): TableWinds {
            return when (seatPosition) {
                EAST -> NORTH
                SOUTH -> WEST
                WEST -> EAST
                NORTH -> SOUTH
                else -> SOUTH
            }
        }
    }
}
